RecoLocalTracker/SiPixelClusterizer/plugins/alpaka/SiPixelPhase2DigiToCluster.cc

```diff
@@ -0,0 +1,167 @@
+    void produce(device::Event& iEvent, device::EventSetup const& iSetup) override;
+
+    const edm::ESGetToken<TrackerGeometry, TrackerDigiGeometryRecord> geomToken_;
+    const edm::EDGetTokenT<edm::DetSetVector<PixelDigi>> pixelDigiToken_;
+
+    device::EDPutToken<SiPixelDigisSoACollection> digiPutToken_;
+    device::EDPutToken<SiPixelDigiErrorsSoACollection> digiErrorPutToken_;
+    device::EDPutToken<SiPixelClustersSoACollection> clusterPutToken_;
+
+    Algo Algo_;
+
+    const bool includeErrors_;
+    const SiPixelClusterThresholds clusterThresholds_;
+    uint32_t nDigis_ = 0;
+
+    SiPixelDigisSoACollection digis_d;
+  };
+
+  SiPixelPhase2DigiToCluster::SiPixelPhase2DigiToCluster(const edm::ParameterSet& iConfig)
+      : geomToken_(esConsumes()),
+        pixelDigiToken_(consumes<edm::DetSetVector<PixelDigi>>(iConfig.getParameter<edm::InputTag>("InputDigis"))),
+        digiPutToken_(produces()),
+        clusterPutToken_(produces()),
+        includeErrors_(iConfig.getParameter<bool>("IncludeErrors")),
+        clusterThresholds_{iConfig.getParameter<int32_t>("clusterThreshold_layer1"),
+                           iConfig.getParameter<int32_t>("clusterThreshold_otherLayers"),
+                           static_cast<float>(iConfig.getParameter<double>("ElectronPerADCGain")),
+                           static_cast<int8_t>(iConfig.getParameter<int>("Phase2ReadoutMode")),
+                           static_cast<uint16_t>(iConfig.getParameter<uint32_t>("Phase2DigiBaseline")),
+                           static_cast<uint8_t>(iConfig.getParameter<uint32_t>("Phase2KinkADC"))} {
+    if (includeErrors_) {
+      digiErrorPutToken_ = produces();
+    }
+  }
+
+  void SiPixelPhase2DigiToCluster::fillDescriptions(edm::ConfigurationDescriptions& descriptions) {
+    edm::ParameterSetDescription desc;
+
+    desc.add<bool>("IncludeErrors", true);
+    desc.add<int32_t>("clusterThreshold_layer1",
+                      pixelClustering::clusterThresholdPhase2LayerOne);  //FIXME (fix the CUDA)
+    desc.add<int32_t>("clusterThreshold_otherLayers", pixelClustering::clusterThresholdPhase2OtherLayers);
+    desc.add<double>("ElectronPerADCGain", 1500.);
+    desc.add<int32_t>("Phase2ReadoutMode", 3);
+    desc.add<uint32_t>("Phase2DigiBaseline", 1000);
+    desc.add<uint32_t>("Phase2KinkADC", 8);
+    desc.add<edm::InputTag>("InputDigis", edm::InputTag("simSiPixelDigis:Pixel"));
+    descriptions.addWithDefaultLabel(desc);
+  }
+
+  void SiPixelPhase2DigiToCluster::acquire(device::Event const& iEvent, device::EventSetup const& iSetup) {
+    auto const& input = iEvent.get(pixelDigiToken_);
+
+    const TrackerGeometry* geom_ = &iSetup.getData(geomToken_);
+
+    uint32_t nDigis = 0;
+    for (auto DSViter = input.begin(); DSViter != input.end(); DSViter++) {
+      nDigis = nDigis + DSViter->size();
+    }
+    SiPixelDigisHost digis_h(nDigis, iEvent.queue());
```

Comment:
Maybe move this line below the `nDigis_ == 0` check to avoid constructing it when not needed?